### PR TITLE
bears/general: Add IndentationBear

### PR DESCRIPTION
**Pull request by abhsag24**
It is a generic indent bear, which looks for a start and end
indent eg: { : }  if they are specified.
eg: "{"  where { is an
indent start.

it does not however support hanging indents or absolute indents of
any sort, also indents based on keywords are not supported yet.
for example:

    if(something)
    does not get indented

changes to:

    if(something)
    does not get indented

Fixes https://github.com/coala-analyzer/coala-bears/issues/40

### CODE REVIEW

**Comment by abhsag24** [comment]
@fneu  review needed :)

File: bears/general/IndentationBear.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+class IndentationBear(LocalBear):
+
+    def run(self,
+            filename,
+            file,
+            dependency_results: dict,
+            language: str,
+            language_family: str,
+            use_tabs_instead_of_spaces: bool=False,
+            indentation_width: int=4):
```
Comment (sils1297) [inline bears/general/IndentationBear.py]: please use the same setting names as the spaceconsistencybear

Comment (sils1297) [inline bears/general/IndentationBear.py]: please use the same setting names as the spaceconsistencybear

File: bears/general/IndentationBear.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+        """
+        This Bear checks and corrects indentation for all languages
+        which provide a mechanism for indentation start and indentation end
+        for eg: ruby has ['do : end'] , C has ['{' : '}'] and so on.
+
+        no support for absolute indentation of any kind or keyword indents yet.
```
Comment (sils1297) [inline bears/general/IndentationBear.py]: elaborate please, this isn't helpful to the user. The commit message is better

File: bears/general/IndentationBear.py (hunk 1 above)
Comment (abhsag24) [inline bears/general/IndentationBear.py]: I have this name in the `RuboCopBear` too we should document these findings so that they can be solved at once

File: bears/general/IndentationBear.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+            Only Applicable if use_tabs_instead_of_spaces is False
+        """
+        lang_settings_dict = LanguageDefinition(language, language_family)
+        annotation_dict = dependency_results[AnnotationBear.name][0].contents
+        indent_types = dict(lang_settings_dict["indent_types"])
+        indent_levels = self.get_indent_levels(
+                                file, filename, indent_types, annotation_dict)
+
+        # This happens only in case of unmatched indents.
+        if type(indent_levels) is type(''):
```
Comment (sils1297) [inline bears/general/IndentationBear.py]: `type('')` is just `str` right?

File: bears/general/IndentationBear.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+            yield Result(
+                      self,
+                      indent_levels,
+                      severity=RESULT_SEVERITY.MAJOR)
+            return
+
+        if use_spaces:
+            insert = ' '*tab_width
+        else:
+            insert = '\t'
```
Comment (sils1297) [inline bears/general/IndentationBear.py]: `insert = ' '*tab_width if use_spaces else '\t'`

File: bears/general/IndentationBear.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+                        close_pos.remove(close_pos[close_index])
+                        open_pos.remove(open_pos[to_match])
+                        to_match -= 1
+                        break
+                    close_index += 1
+                if((len(close_pos) == 0 and to_match != -1) or
+                   (len(close_pos) != 0 and to_match == -1)):
+                    # None to specify unmatched indents
+                    return "Unmatched " + open_indent + ", " +\
+                           close_indent + " pair"
```
Comment (sils1297) [inline bears/general/IndentationBear.py]: it's very weird to retun a str in one situation and another thing in the other and then use it as an error message. You could just raise an exception here which would make more sense I think.

File: bears/general/IndentationBear.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+
+        # Ranges are returned in the order of least nested to most nested
+        # and also on the basis of which come first
+        return tuple(ranges)[::-1]
+
+    @staticmethod
+    def get_valid_chars(file, char, annotation_dict):
+        """
+        A vaild char is a char that is outside of comments or strings.
+
```
Comment (sils1297) [inline bears/general/IndentationBear.py]: param file missing

File: bears/general/IndentationBear.py [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+        :param char:            Char whose validity is to be checked.
+        :param annotation_dict: A dictionary containing sourceranges of all the
+                                strings and comments within a file.
```
Comment (sils1297) [inline bears/general/IndentationBear.py]: return missing and not obvious

File: bears/general/IndentationBear.py (hunk 5 above)
Comment (abhsag24) [inline bears/general/IndentationBear.py]: sure :)

File: bears/general/IndentationBear.py (hunk 3 above)
Comment (Makman2) [inline bears/general/IndentationBear.py]: why not using `instanceof`?

Comment (sils1297) [inline bears/general/IndentationBear.py]: type comparision is something else FWIW @Makman2 !

Sincerely,

Lasse Schuirmann

lasse@schuirmann.net
http://coala-analyzer.org/ - http://viperdev.io/ - http://gitmate.com/

**Comment by sils1297** [comment]
marking wip as per review

File: bears/general/IndentationBear.py (hunk 6 above)
Comment (abhsag24) [inline bears/general/IndentationBear.py]: do we really need it? it gets a bit redundant

File: bears/general/IndentationBear.py [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+                   (len(close_pos) != 0 and to_match == -1)):
+                    # None to specify unmatched indents
+                    raise UnmatchedIndentError(open_indent, close_indent)
+
+        # Ranges are returned in the order of least nested to most nested
+        # and also on the basis of which come first
+        return tuple(ranges)[::-1]
+
+    @staticmethod
+    def get_valid_sequences(file, sequence, annotation_dict):
```
Comment (abhsag24) [inline bears/general/IndentationBear.py]: @sils1297  do we really need a file param? it gets reduntant

**Comment by abhsag24** [comment]
@fneu review needed again :P

File: bears/general/IndentationBear.py (hunk 8 above)
Comment (sils1297) [inline bears/general/IndentationBear.py]: content, stream, name, list of lines or joined string. Docs should explain that.

**Comment by fneu** [comment]
>    if(something)
    does not get indented

>changes to:

>    if(something)
    does not get indented

huh? So does it change or not? If it doesn't change why bother writing the same thing twice?

**Comment by fneu** [comment]
> Fixes #40

Hm, are you sure this is complete enough to fulfill #40? 
Right before that you listed valuable stuff that is no supported yet.

**Comment by fneu** [comment]
close -> misclick, sorry

File: bears/general/IndentationBear.py [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+        """
+        Gets the level of indentation of each line.
+
+        :param indent_types:    A dictionary with keys as start of indent and
+                                values as their corresponding closing indents.
+        :param annotation_dict: A dictionary containing sourceranges of all the
+                                strings and comments within a file.
+        :return:                A tuple containing the levels of indentation of
+                                each line.
+        """
```
Comment (fneu) [inline bears/general/IndentationBear.py]: same thing as sils1297 noted further down, please add a description of *all* arguments to the docstring. I might be repetitive but consistency is a valuable good.

File: bears/general/IndentationBear.py [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+        ranges = self.get_indentation_sourcerange(
+                         file, filename, indent_types, annotation_dict)
+        indent, next_indent = 0, 0
+        for line in range(0, len(file)):
+            indent = next_indent
+            for _range in ranges:
+                if _range.start.line == line + 1:
+                    next_indent += 1
+                if(_range.end.line == line + 1 and
+                   (file[line].lstrip()[0] in indent_types.values())):
```
Comment (fneu) [inline bears/general/IndentationBear.py]: so

```
if (something){
    code
}
```

and

```
if (something){
    code
    }
```

are fine but 

```
if (something){
    code}
```

is not, hm?
well it's probably a bad idea anyways

File: bears/general/IndentationBear.py [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+
+        return tuple(indent_levels)
+
+    def get_indentation_sourcerange(self,
+                                    file,
+                                    filename,
+                                    indent_types,
+                                    annotation_dict):
+        """
+        Gets a sourceranges of all the indentation blocks present inside the
```
Comment (fneu) [inline bears/general/IndentationBear.py]: docstring again.

at least for the "file" param it's actually kinda interesting in which form you expect that.

File: bears/general/IndentationBear.py [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+                                the range of the outermost indentation while
+                                last being the range of the most
+                                nested/innermost indentation.
+                                Equal level indents appear in the order of
+                                first encounter or left to right.
+        """
+        ranges = []
+        for open_indent in indent_types:
+            close_indent = indent_types[open_indent]
+            open_pos = list(IndentationBear.get_valid_sequences(
```
Comment (fneu) [inline bears/general/IndentationBear.py]: this is not a static function so why would you not use `self`?

**Comment by fneu** [comment]
all in all this looks solid but rather basic. I'd very much advise you to take a throughout look at [reindent.py](http://svn.python.org/view/python/trunk/Tools/scripts/reindent.py?view=markup).

It surely is only tangentially related to this because it is python specific and does pretty much everything but key-char-based indenting, but maybe it can give you inspiration on how to improve this bear exactly because of this. Also it's conveniently written in python itself.

File: bears/general/IndentationBear.py (hunk 10 above)
Comment (abhsag24) [inline bears/general/IndentationBear.py]: its actually like 
```
if(something){
    code
}
```
and 
```
if(something){
    code}
```

are fine but 

```
if(something){
    code
    }
``` 
isn't it's actually loosely based on the [K&R](https://en.wikipedia.org/wiki/Indent_style#Tabs.2C_spaces.2C_and_size_of_indent) style indent 
though i plan to have it configurable later on

Comment (fneu) [inline bears/general/IndentationBear.py]: Ah ok. Well the first one is the important one :)

**Comment by abhsag24** [comment]
@fneu review required again

File: bears/general/IndentationBear.py (hunk 9 above)
Comment (abhsag24) [inline bears/general/IndentationBear.py]: added

File: bears/general/IndentationBear.py (hunk 11 above)
Comment (abhsag24) [inline bears/general/IndentationBear.py]: added

**Comment by abhsag24** [comment]
@fneu review pending

File: bears/general/IndentationBear.py [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+            Name of the file that needs to be checked.
+        :param language_family:
+            Language family to be used for indentation.
+        :param language:
+            Language to be used for indentation.
+        :param use_spaces:
+            Insert spaces instead of tabs for indentation.
+        :param tab_width:
+            No. of spaces to insert for indentation.
+            Only Applicable if use_spaces is False
```
Comment (fneu) [inline bears/general/IndentationBear.py]: period missing?

also:

- please put the arguments in the same order in that they appear in the function definition above (line 15 - 21).
- please mention the depencendy_results argument

File: bears/general/IndentationBear.py [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+        for sequence_match in unescaped_search_for(sequence, file_string):
+            valid = True
+            sequence_position = AbsolutePosition(
+                                    file, sequence_match.start())
+            # ignore if within string
+            for string in annotation_dict['strings']:
+                if(sequence_position >= string.start and
+                   sequence_position <= string.end):
+                    valid = False
+            # ignore if within comments
```
Comment (fneu) [inline bears/general/IndentationBear.py]: generally, don't be afraid of newlines ;)

they are in no way required but personally I would put blank lines in front of lines 187, 191 and 196. Just look though your code where else they look nice and try to adapt it as a habit :)

It makes the code easier to read and nobody cares about number of lines of a module.

File: tests/general/IndentationBearTest.py [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+from coalib.settings.Section import Section
+from coalib.settings.Setting import Setting
+
+
+class IndentationBearTest(unittest.TestCase):
+
+    def setUp(self):
+        self.section = Section("")
+        self.section.append(Setting('language', 'c'))
+        self.section.append(Setting('language_family', 'c'))
```
Comment (fneu) [inline tests/general/IndentationBearTest.py]: if we push your other commit first, this won't be needed, right?

File: tests/general/IndentationBearTest.py [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+        invalid_file2 =\
+            ("a {\n",
+             "\tindentlevel1;\n",
+             "\t\tsecondlinehere;\n",
+             "}\n")
+        self.verify_bear(valid_file2, invalid_file2)
+
+    def test_within_strings(self):
+        valid_file1 =\
+            ('"indent withinstring{"\n',
```
Comment (fneu) [inline tests/general/IndentationBearTest.py]: I'm kind of unhappy with your use of the word "indent", but also maybe I'm wrong.

I understand that an indent is the whitespace in front of a line of code, while `{` is merely a symbol that indicates that an indent is needed for the following line. If this is so I'd like to ask you to search for your use of `"indent"` in this file and rephrase where appropriate (including the commit and PR messages).

File: tests/general/IndentationBearTest.py (hunk 15 above)
Comment (abhsag24) [inline tests/general/IndentationBearTest.py]: yes :)

File: tests/general/IndentationBearTest.py [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+                        '}\n')
+        self.verify_bear(valid_file, invalid_file, section)
+
+    def test_unmatched_indents(self):
+        valid_file = ('{}\n',)
+        invalid_file = ('{\n',)
+        self.verify_bear(valid_file, invalid_file)
+
+        invalid_file2 = ('{}}\n',)
+        self.verify_bear(valid_file=None, invalid_file=invalid_file2)
```
Comment (fneu) [inline tests/general/IndentationBearTest.py]: the tests look pretty good. did you try this with a longer piece of real code, like a >100 lines file of c code? Just to see whether it throws up false errors and in what cases it does so.

You don't encessarily need to fix all occursions of such errors yet, but as with hanging indent please note the limitation in the commit message :)

Comment (abhsag24) [inline tests/general/IndentationBearTest.py]: i tried with the files of the git repository , some with 2000-3000 lines, so the basic problem is the keywords but other than it works!

**Comment by abhsag24** [comment]
@fneu ping :)

File: bears/general/IndentationBear.py [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+    def run(self,
+            filename,
+            file,
+            dependency_results: dict,
+            language: str,
+            use_spaces: bool=True,
+            tab_width: int=4):
+        """
+        It is a generic indent bear, which looks for a start and end
+        indent, example: ``{ : }`` where "{" is the start indent and
```
Comment (fneu) [inline bears/general/IndentationBear.py]: also indent spedifier ;)

File: bears/general/IndentationBear.py [hunk 19]
```diff
@@ -0,0 +1,218 @@
+from coalib.bears.LocalBear import LocalBear
+from bears.general.AnnotationBear import AnnotationBear
+from bears.general.AnnotationBear import starts_within_ranges
+from coalib.parsing.StringProcessing.Core import unescaped_search_for
+from coalib.bearlib.languages.LanguageDefinition import LanguageDefinition
+from coalib.results.SourceRange import SourceRange
+from coalib.results.Result import Result, RESULT_SEVERITY
+from coalib.results.AbsolutePosition import AbsolutePosition
+from coalib.results.Diff import Diff
```
Comment (fneu) [inline bears/general/IndentationBear.py]: what happened to ordering them alphabetically?

actually, looking at [PEP8](https://www.python.org/dev/peps/pep-0008/#imports) again:

> Imports should be grouped in the following order:
> - standard library imports
> - related third party imports
> - local application/library specific imports
>
> You should put a blank line between each group of imports. 

so at least please put bear imports at the top :)

**Comment by fneu** [comment]
@abhsag24 2 minor things you missed, fine besides.

File: bears/general/IndentationBear.py (hunk 19 above)
Comment (abhsag24) [inline bears/general/IndentationBear.py]: or rather te bear imports on the botto right? those are the local ones

**Comment by sils1297** [comment]
@abhsag24 you aware of failing coverage?

**Comment by abhsag24** [comment]
@sils1297 yes, i pushed for @fneu to see the algorithms i'll mark it as wip

File: bears/general/IndentationBear.py (hunk 19 above)
Comment (fneu) [inline bears/general/IndentationBear.py]: yes you are right

**Comment by fneu** [comment]
ack https://github.com/coala-analyzer/coala-bears/pull/468/commits/f39ed92a39952c0308d970370883b7002fc3767a

**Comment by abhsag24** [comment]
@fneu need review!

File: bears/general/AnnotationBear.py [hunk 20]
```diff
@@ -9,21 +9,22 @@
 
 class AnnotationBear(LocalBear):
 
-    def run(self, filename, file, language: str):
+    def run(self, filename, file, language: str, coalang_dir: str = None):
```
Comment (fneu) [inline bears/general/AnnotationBear.py]: does this commit even fit into the PR "Add Intendation bear"?

It's not a problem since in the end its all about the commit but if this is independent we can probably just merge it in an own PR.

File: requirements.txt [hunk 21]
```diff
@@ -1,5 +1,5 @@
 # Use >= for development versions so that source builds always work
-coala>=0.7.0.dev20160602092230
+coala>=0.7.0.dev20160605225837
```
Comment (fneu) [inline requirements.txt]: this is related and up to date?

File: bears/general/IndentationBear.py [hunk 22]
```diff
@@ -47,7 +48,8 @@ def run(self,
             No. of spaces to insert for indentation.
             Only Applicable if use_spaces is False.
         """
```
Comment (fneu) [inline bears/general/IndentationBear.py]: please again add a description of the added param `coalang_dir`

File: bears/general/AnnotationBear.py (hunk 20 above)
Comment (abhsag24) [inline bears/general/AnnotationBear.py]: yeah i guess you're right i'll put this in an own PR

File: requirements.txt (hunk 21 above)
Comment (abhsag24) [inline requirements.txt]: i'm not sure if its up do date but yes its related

File: bears/general/IndentationBear.py (hunk 22 above)
Comment (abhsag24) [inline bears/general/IndentationBear.py]: yes sorry i tend to miss those often, we need a bear which tells us that :(

Comment (sils1297) [inline bears/general/IndentationBear.py]: Ping @sanketdg about that

Comment (abhsag24) [inline bears/general/IndentationBear.py]: there's already an issue in coala-bears regarding this IIRC i'll assign him to that :P

File: bears/general/IndentationBear.py [hunk 23]
```diff
@@ -51,13 +53,39 @@ def run(self,
         lang_settings_dict = LanguageDefinition(
                             language, coalang_dir=coalang_dir)
         annotation_dict = dependency_results[AnnotationBear.name][0].contents
-        indent_types = dict(lang_settings_dict["indent_types"])
+        # sometimes can't convert strings with ':' to dict correctly
+        # as a work around it is required in the specification to always put the
+        # colon last
```
Comment (fneu) [inline bears/general/IndentationBear.py]: we talked about a work around using lists that allow us to keep the colon wherever we want

File: bears/general/IndentationBear.py [hunk 24]
```diff
@@ -51,13 +53,39 @@ def run(self,
...
+        indent_types = lang_settings_dict["indent_types"]
+        if (':' in str(indent_types).rstrip()[-1] and
+                ":" not in dict(indent_types).keys()):
+            indent_types = dict(indent_types)
+            indent_types[":"] = ""
+        else:
+            indent_types = dict(indent_types)
+
+        encapsulators = (dict(lang_settings_dict["encapsulators"]) if
```
Comment (fneu) [inline bears/general/IndentationBear.py]: anybody has suggestions for a better name?

**Comment by fneu** [comment]
```
def func(x):
    a = [1,2,
3,4]
```

needs to be supported, i.e. not handled as an unindent

**Comment by abhsag24** [comment]
reack 11e9b24

**Comment by abhsag24** [comment]
@fneu ping :)

File: bears/general/IndentationBear.py [hunk 25]
```diff
@@ -46,8 +47,12 @@ def run(self,
         :param tab_width:
             No. of spaces to insert for indentation.
             Only Applicable if use_spaces is False.
+        :param coalang_dir:
+            Full path of external directory containing the coalang
+            file for language.
         """
-        lang_settings_dict = LanguageDefinition(language)
+        lang_settings_dict = LanguageDefinition(
+                            language, coalang_dir=coalang_dir)
```
Comment (fneu) [inline bears/general/IndentationBear.py]: up to here, all of this is coalang related, right? could that be an own commit similar to https://github.com/coala-analyzer/coala-bears/commit/9542e176071cb64a4f7676f98d09609c64bb3245 maybe?

File: bears/general/IndentationBear.py [hunk 26]
```diff
@@ -60,14 +65,13 @@ def run(self,
             return
 
         insert = ' '*tab_width if use_spaces else '\t'
+        no_indent_file = [line.lstrip() if line.lstrip() is not "" else "\n"
+                          for line_nr, line in enumerate(file)]
```
Comment (fneu) [inline bears/general/IndentationBear.py]: all strings besides empty strings have a boolean value of `True`, right?

so you could write:
```python
no_indent_file = [line.lstrip() if line.lstrip() else "\n"
                          for line_nr, line in enumerate(file)]
```

I'm not always a fan of writing everything as compact as possible but here I feel like that makes it a little less convoluted.

File: bears/general/IndentationBear.py [hunk 27]
```diff
@@ -60,14 +65,13 @@ def run(self,
...
-        for line_nr, line in enumerate(file):
-            # Leave out empty lines
-            if line.lstrip() != '':
-                new_file.append(insert*indent_levels[line_nr] +
-                                line.lstrip())
-            else:
-                new_file.append('\n')
+        for line_nr, line in enumerate(no_indent_file):
+            new_file.append(insert*indent_levels[line_nr] +
+                            line if line is not "\n" else "\n")
```
Comment (fneu) [inline bears/general/IndentationBear.py]: The line break confuses me. It looks like you add the indents and then (in the next line) either the line or a new line.

so at least please put the first `line` into the line above and maybe a line break in front of the `else`?
just something that underlines your intent

File: bears/general/IndentationBear.py [hunk 28]
```diff
@@ -93,19 +97,29 @@ def get_indent_levels(self, file, filename, indent_types, annotation_dict):
...
             indent = next_indent
             for _range in ranges:
                 if _range.start.line == line + 1:
                     next_indent += 1
+
                 if(_range.end.line == line + 1 and
                    (file[line].lstrip()[0] in indent_types.values())):
                     indent -= 1
                     next_indent -= 1
+
```
Comment (fneu) [inline bears/general/IndentationBear.py]: empty lines! yeay, love them :smile: 

no, really.

but baaaasically, if you would insert new lines in an own commit, I would just ack that very quickly and we were done with it. This way it kinda bloats your commit.

You don't need to remove them now from this commit, but generally, splitting your changes into a lot of atomic commits can make things easier and simpler. Like with the coalang stuff at the top of this file.

I'm not afraid to review a PR with lot's of commits... I'm afraid of a PR with gigantic commits that take an hour to review each :smile: 

your new lines here are not that bad, but it needed to be told anyways :D

File: bears/general/IndentationBear.py [hunk 29]
```diff
@@ -136,33 +151,32 @@ def get_specified_block_range(self,
                                 first encounter or left to right.
         """
         ranges = []
-        for open_indent in indent_types:
-            close_indent = indent_types[open_indent]
-            open_pos = list(self.get_valid_sequences(
+
+        open_pos = list(self.get_valid_sequences(
```
Comment (fneu) [inline bears/general/IndentationBear.py]: just noted this here, might be relevant in other places: 

quoting PEP8:

```python
# Aligned with opening delimiter.
foo = long_function_name(var_one, var_two,
                         var_three, var_four)

# More indentation included to distinguish this from the rest.
def long_function_name(
        var_one, var_two, var_three,
        var_four):
    print(var_one)

# Hanging indents should add a level.
foo = long_function_name(
    var_one, var_two,
    var_three, var_four)
```

you sometimes break a line like here and then indent it way to much. We don't have an indent bear yet, look out for this :smile:

File: bears/general/IndentationBear.py [hunk 30]
```diff
@@ -136,33 +151,32 @@ def get_specified_block_range(self,
...
                                      file, open_indent, annotation_dict))
-            close_pos = list(self.get_valid_sequences(
+        close_pos = list(self.get_valid_sequences(
                                      file, close_indent, annotation_dict))
```
Comment (fneu) [inline bears/general/IndentationBear.py]: also here

File: bears/general/IndentationBear.py [hunk 31]
```diff
@@ -136,33 +151,32 @@ def get_specified_block_range(self,
...
-                    # None to specify unmatched indents
-                    raise UnmatchedIndentError(open_indent, close_indent)
+        to_match = len(open_pos) - 1
+        while to_match >= 0:
+            close_index = 0
+            while close_index < len(close_pos):
+                if(open_pos[to_match].position
+                   <= close_pos[close_index].position):
+                    ranges.append(
+                        SourceRange.from_absolute_position(
```
Comment (fneu) [inline bears/general/IndentationBear.py]: and here

File: tests/general/IndentationBearTest.py [hunk 32]
```diff
@@ -1,18 +1,22 @@
...
 
 
 class IndentationBearTest(unittest.TestCase):
 
     def setUp(self):
         self.section = Section("")
-        self.section.append(Setting('language', 'c'))
+        self.section.append(Setting('language', 'test'))
         self.section.append(Setting('use_spaces', False))
+        self.section.append(Setting('coalang_dir', escape(os.path.join(
```
Comment (fneu) [inline tests/general/IndentationBearTest.py]: here

File: tests/general/IndentationBearTest.py [hunk 33]
```diff
@@ -151,3 +155,14 @@ def test_unmatched_indents(self):
 
         invalid_file2 = ('{}}\n',)
         self.verify_bear(valid_file=None, invalid_file=invalid_file2)
+
+    def test_multiple_indent_specifiers(self):
```
Comment (fneu) [inline tests/general/IndentationBearTest.py]: I made this line yellow, I don't know how :D

anyway, have you though about using multi-line strings like so for your test?

```python
valid file = """
multi
line
    indented
file
""".splitlines(True)[1:]
```

or maybe just do the same you are doing now but indent with spaces instead of tabs? That would make it a little easier to read which lines are indented

File: bears/general/IndentationBear.py (hunk 25 above)
Comment (abhsag24) [inline bears/general/IndentationBear.py]: i thought the external coalang helps in the testing of the commit so it might be logical that it can go in this commit? i have nothing against it going into a new commit though :P

File: bears/general/IndentationBear.py [hunk 34]
```diff
@@ -54,13 +56,40 @@ def run(self,
...
+        else:
+            indent_types = dict(lang_settings_dict["indent_types"])
+
+        encapsulators = (dict(lang_settings_dict["encapsulators"]) if
+                         "encapsulators" in lang_settings_dict else {})
+
+        encaps_pos = []
+        for encapsulator in encapsulators:
+            encaps_pos += self.get_specified_block_range(
+                                   file,
```
Comment (fneu) [inline bears/general/IndentationBear.py]: weird indentation again, I would use:
```python
get_specified_block_range(file,
                          filename,
                          encapsulator,
                          [...]
```

Comment (fneu) [inline bears/general/IndentationBear.py]: also, I'm just noticing this now because of the syntax highlight in the comment above:

`file` is a really bad name for a variable if it overwrites a builtin... I know it does so in Python2, unsure about Python3, could you please check?

also, if you refactor this, please in an own commit, ok?

File: bears/general/IndentationBear.py [hunk 35]
```diff
@@ -176,12 +226,57 @@ def get_specified_block_range(self,
...
+        :param comments:        A dict containing all the types of comments
+                                specifiers in a language.
+        """
+        specifiers = list(self.get_valid_sequences(
+                                   file, indent_specifier, annotation_dict))
+        _range = []
+        for specifier in specifiers:
+            current_line = specifier.line
+            indent = get_indent_of_specifier(file, current_line, encapsulators)
+            unindent_line = get_first_unindent(indent,
```
Comment (fneu) [inline bears/general/IndentationBear.py]: good intendation :)

File: bears/general/IndentationBear.py [hunk 36]
```diff
@@ -176,12 +226,57 @@ def get_specified_block_range(self,
...
+                                               file,
+                                               current_line,
+                                               annotation_dict,
+                                               encapsulators,
+                                               comments)
+
+            if unindent_line == specifier.line:
+                raise ExpectedIndentError(specifier.line)
+
+            _range.append(SourceRange.from_values(
```
Comment (fneu) [inline bears/general/IndentationBear.py]: bad intendation :(

File: tests/general/IndentationBearTest.py [hunk 37]
```diff
@@ -166,3 +166,47 @@ def test_multiple_indent_specifiers(self):
                         '\t<\n',
                         '\t not giving indentation>}\n')
         self.verify_bear(valid_file, invalid_file)
+
+    def test_unspecified_unindents(self):
+        valid_file = ('switch(expression) {\n',
```
Comment (fneu) [inline tests/general/IndentationBearTest.py]: a little hard to read again

File: bears/general/IndentationBear.py (hunk 25 above)
Comment (fneu) [inline bears/general/IndentationBear.py]: well, if you add that commit before this one, then it still helps with 
testing, not?

File: tests/general/IndentationBearTest.py (hunk 33 above)
Comment (abhsag24) [inline tests/general/IndentationBearTest.py]: SpaceConsistencyBear likes to eat up those extra leading spaces in my strings, that's why i use, tabs and only use spaces when necessary

Comment (fneu) [inline tests/general/IndentationBearTest.py]: hahaha wow ok. coala is too good to be cheated.

well I guess it's fine

File: bears/general/IndentationBear.py (hunk 34 above)
Comment (abhsag24) [inline bears/general/IndentationBear.py]: file is not a builtin in python3, also we require it to be named file right? its the case with all the bears

Comment (fneu) [inline bears/general/IndentationBear.py]: inside the function you can call it whatever, but if that is not doing any harm in python3 we are good anyways.

**Comment by jschwarzwalder** [comment]
Will this work with html and css?

**Comment by abhsag24** [comment]
css yes, html not tat the moment, though that's not too tough, i think, i might be able to do that as well

File: bears/general/IndentationBear.py (hunk 34 above)
Comment (abhsag24) [inline bears/general/IndentationBear.py]: also i cant align like that because of the line length, though i will correct the indentation

**Comment by abhsag24** [comment]
reack 3ebc5e5

File: bears/general/IndentationBear.py (hunk 25 above)
Comment (fneu) [inline bears/general/IndentationBear.py]: indentation ;)

File: bears/general/IndentationBear.py [hunk 38]
```diff
@@ -98,19 +97,29 @@ def get_indent_levels(self, file, filename, indent_types, annotation_dict):
         :return:                A tuple containing the levels of indentation of
                                 each line.
         """
+        ranges = []
+        for indent_specifier in indent_types:
+            ranges += self.get_specified_block_range(
```
Comment (fneu) [inline bears/general/IndentationBear.py]: indentation

File: bears/general/IndentationBear.py [hunk 39]
```diff
@@ -54,13 +56,38 @@ def run(self,
...
+        comments = dict(lang_settings_dict["comment_delimiter"])
+        comments.update(
+            dict(lang_settings_dict["multiline_comment_delimiters"]))
 
         try:
             indent_levels = self.get_indent_levels(
-                                  file, filename, indent_types, annotation_dict)
-        # This happens only in case of unmatched indents.
-        except UnmatchedIndentError as e:
+                                    file,
```
Comment (fneu) [inline bears/general/IndentationBear.py]: indentation

File: bears/general/IndentationBear.py [hunk 40]
```diff
@@ -94,17 +127,30 @@ def get_indent_levels(self, file, filename, indent_types, annotation_dict):
...
+                                specifiers in a language.
         :return:                A tuple containing the levels of indentation of
                                 each line.
         """
         ranges = []
         for indent_specifier in indent_types:
-            ranges += self.get_specified_block_range(
+            if indent_types[indent_specifier]:
+                ranges += self.get_specified_block_range(
                                     file,
```
Comment (fneu) [inline bears/general/IndentationBear.py]: indentation

File: bears/general/IndentationBear.py [hunk 41]
```diff
@@ -94,17 +127,30 @@ def get_indent_levels(self, file, filename, indent_types, annotation_dict):
...
                                     filename,
                                     indent_specifier,
                                     indent_types[indent_specifier],
                                     annotation_dict)
+            else:
+                ranges += self.get_unspecified_block_range(
+                                    file,
```
Comment (fneu) [inline bears/general/IndentationBear.py]: indentation

**Comment by fneu** [comment]
```python
for line in commit:
    if "file" in line:
        watch(indentation)
```

**Comment by fneu** [comment]
really good besides, I think it's merge-able on time for the midterm :)

**Comment by fneu** [comment]
ack https://github.com/coala-analyzer/coala-bears/pull/468/commits/3b2ec868c97c9b672f01b8295f9348f8218a3603 https://github.com/coala-analyzer/coala-bears/pull/468/commits/4ec5c101099237dd985121845e5f1fd9dc1d3def https://github.com/coala-analyzer/coala-bears/pull/468/commits/caaa8f2d4940e774f66983c197d60e6998c5345d

**Comment by fneu** [comment]
@rultor merge this, please

**Comment by fneu** [comment]
@rultor merge now!
